crates/extensions/c8y_firmware_manager/src/message.rs

```diff
@@ -0,0 +1,312 @@
+use c8y_api::smartrest::smartrest_serializer::TryIntoOperationStatusMessage;
+use mqtt_channel::Topic;
+use tedge_api::OperationStatus;
+use tedge_mqtt_ext::MqttMessage;
+
+#[derive(Debug)]
+pub struct FirmwareOperationRequest {
+    child_id: String,
+    payload: RequestPayload,
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Debug)]
+pub struct RequestPayload {
+    #[serde(rename = "id")]
+    pub operation_id: String,
+    pub attempt: usize,
+    pub name: String,
+    pub version: String,
+    pub sha256: String,
+    #[serde(rename = "url")]
+    pub file_transfer_url: String,
+}
+
+impl FirmwareOperationRequest {
+    fn get_topic(&self) -> Topic {
+        Topic::new_unchecked(&format!(
+            "tedge/{}/commands/req/firmware_update",
+            self.child_id
+        ))
+    }
+
+    fn get_json_payload(&self) -> Result<String, serde_json::Error> {
+        serde_json::to_string(&self.payload)
+    }
+}
+
+impl From<FirmwareOperationEntry> for FirmwareOperationRequest {
+    fn from(entry: FirmwareOperationEntry) -> Self {
+        Self {
+            child_id: entry.child_id.to_string(),
+            payload: RequestPayload {
+                operation_id: entry.operation_id.to_string(),
+                attempt: entry.attempt,
+                name: entry.name.to_string(),
+                version: entry.version.to_string(),
+                sha256: entry.sha256.to_string(),
+                file_transfer_url: entry.file_transfer_url,
+            },
+        }
+    }
+}
+
+impl TryInto<MqttMessage> for FirmwareOperationRequest {
+    type Error = FirmwareManagementError;
+
+    fn try_into(self) -> Result<MqttMessage, Self::Error> {
+        let message = MqttMessage::new(&self.get_topic(), self.get_json_payload()?);
+        Ok(message)
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct FirmwareOperationResponse {
+    child_id: String,
+    payload: ResponsePayload,
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Debug, PartialEq, Eq)]
+pub struct ResponsePayload {
+    #[serde(rename = "id")]
+    pub operation_id: String,
+    pub status: OperationStatus,
+    pub reason: Option<String>,
+}
+
+impl FirmwareOperationResponse {
+    pub fn get_child_id(&self) -> String {
+        self.child_id.clone()
+    }
+
+    pub fn get_payload(&self) -> &ResponsePayload {
+        &self.payload
+    }
+}
+
+impl TryFrom<&MqttMessage> for FirmwareOperationResponse {
+    type Error = FirmwareManagementError;
+
+    fn try_from(message: &MqttMessage) -> Result<Self, Self::Error> {
+        let topic = &message.topic.name;
+        let child_id = get_child_id_from_child_topic(topic)?;
+        let request_payload: ResponsePayload = serde_json::from_str(message.payload_str()?)?;
+
+        Ok(Self {
+            child_id,
+            payload: request_payload,
+        })
+    }
+}
+
+// FIXME: Duplicated with config plugin
+pub fn get_child_id_from_child_topic(topic: &str) -> Result<String, FirmwareManagementError> {
```

Comment:
You can move that function to `tedge_api::topic`.

See https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/tedge_api/src/topic.rs#L60